---
title: "The BCB420.2019.ESA Vignette"
author:
- name: Boris Steipe
  affiliation: 
  - http://orcid.org/0000-0002-1134-6758
  - University of Toronto
  email: boris.steipe@utoronto.ca
date: "`r Sys.Date()`"
abstract: >
  This Vignette is a placeholder for global package documentation.
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{BCB420.2019.ESA Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
references:
- id: steipe2019
  title: 'rptPlus: R Package Template - enhanced'
  author:
  - family: Steipe
    given: Boris
    orcid: 0000-0002-1134-6758
  container-title: '[R package](https://github.com/hyginn/BCB420.2019.ESA)'
  URL: 'https://github.com/hyginn/BCB420.2019.ESA'
---

<!--- this inline style supports diff-like highlighting -->
<style>
  .add {
    color: #22863a;
    background-color: #f0fff4;
  }
  .del {
    color: #b31d28;
    background-color: #ffeef0;
  }
</style>


```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

# knitr knits in a new session with an empty global workspace after setting its
# working directory to ./vignettes. To make your package functions available in
# the vignette, you have to load the library. The following two lines should
# accomplish this without manual intervention:
pkgName <- trimws(gsub("^Package:", "", readLines("../DESCRIPTION")[1]))
library(pkgName, character.only = TRUE)

```

&nbsp;

<div style="font-size:90%;border:solid 1px #337ab7;background-color:#FAFBFC;margin-right:0px;">
There are many links and references in this document. If you find anything here
ambiguous, inaccurate, outdated, incomplete, or broken, please
[file an issue](https://github.com/hyginn/BCB420.2019.ESA/issues)!
</div>

&nbsp;

# About this Vignette

This sample Vignette explains the idea behind the forked BCB420.2019.ESA package containing the functions SeqComparisonTable.R and fetchPDBHGNCdatabase.R.

The purpose of SeqComparisonTable is to generate a data frame containing a comparison between the resulting protein sequence of an HGNC gene of interest with it's interacting genes/proteins. The aim is to compare the resulting protein sequences of interacting genes to give insight to the biological properties of encoded genes in a curated system. This is achieved by comparing a gene of interest and it's encoding protein sequence to it's interacting genes/proteins from the STRING database. A resulting data frame contains the gene of interest, the compared gene and it's protein sequences, along with the consensus sequence and conservation scores generated by a ClustalW alignment algorithm and BLOSUM62 scoring matrix. 

The functions defined in this forked repository uses pre-defined functions in the BCB420.2019.ESA package from the master branch. 

## Why is comparison of sequences in a system important? 

It is a neccessity to compare protein sequences to retrieve insight on protein/gene of interest. Conservation analysis and multiple sequence alignments have been proven to be a powerful indicator of functional importance in proteins and genes. Often it is used to detect residues involved in ligand binding, in proteinâ€“protein interaction interfaces, in maintaining structure, and in determining protein functional specificity.Conservation analysis has also been used with structural information for many applications. 

In a systems analysis, conservation analysis of interacting genes and proteins can provide insight on the functional important and relationships between the components of the system. Analysis on the interacting protein sequence can expose interaction interfaces and ligand binding between the proteins and genes. As well, interacting proteins under the same family may have conserved regions in their sequence which can infer a homology and similar evolutionary descent. All in all, sequence comparison and analysis can generate a high degree of information, especially in a system. This tool aims to generate this sequence comparsion.

# Example usage

'''
# Pick a HGNC symbol in the PHALY system as an example for this case.
> hgnc <- "BECN1"

# Call the SeqComparisonTable function to generate the dataframe
# The datafram will contain the corresponding protein alignments
# with it's interacting genes
> seqComparisonDF <- SeqComparisonTable(hgnc)
> (seqComparisonDF)

'''

=======


# On input data and preparation
<a name="a1"></a>
- https://github.com/judyheewonlee/BCB420.2019.PDB The readME file contains further information on the relationships between PDB and HGNC symbols and how they are produced
- https://github.com/hyginn/BCB420.2019.ESA The master repository for the BCB420.2019.ESA package

# Session Info

This release of the `BCB420.2019.ESA` package was produced in the following context of supporting packages:

```{r, echo=TRUE}
sessionInfo()
```

<!-- the bibliography appears at the end of the page -->
# References
- https://www.ncbi.nlm.nih.gov/pubmed/17519246 Discussion on the importance of multiple sequence alignments and conservation analysis
- http://steipe.biochemistry.utoronto.ca/abc/students/index.php/User:Judy_Lee/BCB420-2019-ExploratorySystemsAnalysis Link to the project page for the functions for the BCB420.2019.ESA package


<!-- End -->
