# SeqComparisonTable.R
#
# Purpose: Exploratory System Analysis tool project for BCB420 2019
# Version: 1.0.0
# Version history: 1.0.0
# Date: March 24 2019
# Author: Heewon Lee (heewon.lee@mail.utoronto.ca)
# License: MIT
#
# Input: A HGNC symbol that the user would like to retrieve
# the corresponding protein sequence alignments with it's interacting
# genes according to the STRING database
# Output: A data frame including the hgnc gene of interest,
#'     the gene it is being compared to, the protein sequence of
#'     each resulting protein generated by the genes and
#'     the consensus sequence and conservation scores based
#'     off a ClustalW multiple Sequence Alignment.
# Dependencies: BiocManager, Biostrings, msa

# ====  FUNCTION  =============================================================
# SeqComparisonTableESA.R
#'
#' \code{SeqComparisonTable} Return a data frame containing the alignment data
#'    between the provided hgnc gene and it's corresponding genes based
#'    on the STRING database..
#'
#' @param hgnc (character) The hgnc symbol of the gene of interest in the
#'     form of a string. This hgnc gene will have multiple sequence
#'     alignments being performed on it with all of it's interacting
#'     proteins based off the STRING database.
#'
#' @return (data frame) A data frame including the hgnc gene of interest,
#'     the gene it is being compared to, the protein sequence of
#'     each resulting protein generated by the genes and
#'     the consensus sequence and conservation scores based
#'     off a ClustalW multiple Sequence Alignment.
#'
#' @author {Heewon Lee} (aut)
#' @examples
#' # Retrieve system database using functions provided from the master package
#' myDB <- fetchData("SysDB")
#' # See which genes the "PHALY" systems contains
#' (SyDBgetSysSymbols(myDB, "PHALY"))
#' # Pick a sample HGNC symbol within the PHALY system
#' hgnc <- "BECN1"
#' # Call the SeqComparisonTable function to generate the dataframe
#' seqComparisonDF <- SeqComparisonTable(hgnc)
#'
#' @export

SeqComparisonTable <- function(hgnc) {

  ### =========== Load all required packages and data ================= ###

  # Import STRINGedges and fetchComponents() from git repo
  # from https://github.com/judyheewonlee/BCB420.2019.ESA made
  # by Professor Boris Steipe (forked directory, master repo
  # can be accessed from https://github.com/hyginn/BCB420.2019.ESA)

  # Load STRINGedges
  myURL <- paste0("http://steipe.biochemistry.utoronto.ca/abc/assets/",
                  "STRINGedges-2019-03-14.RData")
  load(url(myURL))  # loads STRING edges object

  if (! requireNamespace("msa", quietly = TRUE)) {
    BiocManager::install("msa")
  }

  if (! requireNamespace("Biostrings", quietly = TRUE)) {
    BiocManager::install("Biostrings")
  }

  ### ============ Setting up PDB-HGNC database ==================== ###

  # Load PDBHGNC data
  PDBHGNCpath <- system.file("extdata", "PDBHGNC.RData",
                             package = "BCB420.2019.ESA")
  load(PDBHGNCpath)

  ### ========== Generate the PDB Sequence comparison Table ======== ###

  # Get the hgnc symbols that interact with the provided hgnc gene
  # using the STRING database
  edges <- STRINGedges[STRINGedges$a == hgnc,]

  hgncTable <- PDBHGNC$Transcripts[PDBHGNC$Transcripts$hgncID
                                   %in% edges$b,c("ID", "hgncID")]

  hgncTable <- rbind(hgncTable, PDBHGNC$Transcripts[PDBHGNC$Transcripts$hgncID
                                              == hgnc, c("ID", "hgncID")])

  hgncTable$Sequence <- NA
  hgncTable$Resolution <- NA


  # Pick the chain with the best resolution if there are overlapping
  # PDB protein structures
  for (transcript in hgncTable$ID) {
    sel <- PDBHGNC$pdbChains[PDBHGNC$pdbChains$transcriptHGNC == transcript,]
    edgeSeq <- sel[sel$Resolution == min(sel$Resolution),
                   ][1,c("Resolution", "Sequences")]
    hgncTable[hgncTable$ID == transcript,]$Sequence <- edgeSeq$Sequences
    hgncTable[hgncTable$ID == transcript,]$Resolution <- edgeSeq$Resolution

  }

  # Perform an MSA using the Bioconductor msa package in order to retrieve
  # the sequence alignments between the provided hgnc protein and
  # all protein chain sequences it interacts with according to the
  # STRING database
  seqComparisonDF <- data.frame()

  tempDF <- data.frame(a = NA,
                       b = NA,
                       SequenceA = NA,
                       SequenceB = NA,
                       Consensus = NA,
                       Score = NA)

  mainTranscript <- hgncTable[hgncTable$hgncID == hgnc,]

  # Compare all protein chains of the gene of interest
  # with all the other protein chains it interacts with
  # using the msa package
  for (a in mainTranscript$ID) {
    tempDF$a <- a
    seqA <- mainTranscript[mainTranscript$ID == a, "Sequence"]

    tempDF$SequenceA <- seqA

    for (b in hgncTable$ID) {
      tempDF$b <- b
      seqB <- hgncTable[hgncTable$ID == b, "Sequence"]

      tempDF$SequenceB <- seqB

      # import BLOSUM62 scoring matrix from Biostrings package from
      # Bioconductor

      data("BLOSUM62", package = "Biostrings")

      myMSA <- msa::msaClustalW(c(seqA, seqB), type="protein")
      myConsensus <- msa::msaConsensusSequence(myMSA)
      myScore <- msa::msaConservationScore(myMSA, BLOSUM62)

      tempDF$Consensus <- myConsensus
      tempDF$Score <- as.list(as.data.frame(myScore))

      seqComparisonDF <- rbind(seqComparisonDF, tempDF)

    }

  }

  return(seqComparisonDF)
}


# [END]
